lib/lhm/sql_retry.rb

```diff
@@ -15,22 +15,109 @@ module Lhm
   # https://github.com/kamui/retriable. Additionally, a "log_prefix" option,
   # which is unique to SqlRetry can be used to prefix log output.
   class SqlRetry
-    def initialize(connection, options = {})
+
+    RECONNECT_SUCCESSFUL_MESSAGE = "LHM successfully reconnected to initial host:"
+    ABNORMAL_EXECUTION_TIME_THRESHOLD = 5
+
+    def initialize(connection, options = {}, with_consistent_host = true, retry_for_host_options = {})
       @connection = connection
       @log_prefix = options.delete(:log_prefix)
       @retry_config = default_retry_config.dup.merge!(options)
+      if (@with_consistent_host = with_consistent_host)
+        @host_retry_config = default_host_retry_config.dup.merge!(retry_for_host_options)
+        @initial_host = hostname
+        @initial_server_id = server_id
+      end
     end
 
-    def with_retries
+    def with_retries(override_prefix=nil)
+      old_prefix =  @log_prefix
+      @log_prefix = override_prefix unless override_prefix.nil?
+
       Retriable.retriable(retry_config) do
-        yield(@connection)
+        if @with_consistent_host
+          raise Lhm::Error.new("Could not reconnected to initial MySQL host. Aborting to avoid data-loss") unless same_host_as_initial?
+        end
+        with_time_check do
+          yield(@connection)
+        end
+      rescue Lhm::Error
+        raise
```

Comment:
Because `Lhm::Error < StandardError` and we want different flows for exception handling. Otherwise, every "internal" error used for control flow would trigger a reconnection and it would loop. I tried removing it to simplify the logic, and it broke it.

Reply:
> I tried removing it to simplify the logic, and it broke it.

What's the second "it"? How did it break?

Also following from my previous comment, can't we use a specific internal error class?